walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> unavailableTimeRanges = getUnavailableTimes(events, request.getAttendees());
+    List<TimeRange> suggestedTimeRanges = findAvailableTimeRanges(unavailableTimeRanges, request.getDuration());
+
+    return suggestedTimeRanges;
+  }
+
+  private List<TimeRange> getUnavailableTimes(Collection<Event> events, Collection<String> attendees) {
+    List<TimeRange> eventTimes = events
+        .stream()
+        .filter(event -> !areAttendeesInEvent(event, attendees))
+        .map(event -> event.getWhen())
+        .sorted(TimeRange.ORDER_BY_START)
+        .collect(Collectors.toList());
+
+    List<TimeRange> unavailableTimes = mergeOverlappingTimeRanges(eventTimes);
+
+    return unavailableTimes;
+  }
+
+  private boolean areAttendeesInEvent(Event event, Collection<String> attendees) {
+    return Collections.disjoint(event.getAttendees(), attendees);
+  }
+
+  private List<TimeRange> mergeOverlappingTimeRanges(List<TimeRange> timeRanges) {
+    ArrayList<TimeRange> mergedTimeRanges = new ArrayList<>();
+
+    for (TimeRange timeRange : timeRanges) {
+      if (mergedTimeRanges.size() == 0) {
+        mergedTimeRanges.add(timeRange);
+        continue;
+      } 
+
+      int lastIndex = mergedTimeRanges.size() - 1;
+      TimeRange lastTimeRange = mergedTimeRanges.get(lastIndex);
+      
+      if (lastTimeRange.contains(timeRange)) {
+        continue;
+      } 
+      
+      if (lastTimeRange.overlaps(timeRange)) {
+        mergedTimeRanges.set(lastIndex, TimeRange.fromStartEnd(lastTimeRange.start(), timeRange.end(), false));
+        continue;
+      } 
+      
+      mergedTimeRanges.add(timeRange);
+    }
+
+    return mergedTimeRanges;
+  }
+
+  private List<TimeRange> findAvailableTimeRanges(List<TimeRange> unavailableTimeRanges, long eventDuration) {
```

Comment:
nit: it's good to include the unit of measurement in the name of `eventDuration` (e.g. `eventDurationMs` or `eventDurationMillis`)